faiss_vector_cache.go

```diff
@@ -0,0 +1,228 @@
+//  Copyright (c) 2024 Couchbase, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 		http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build vectors
+// +build vectors
+
+package zap
+
+import (
+	"sync"
+	"sync/atomic"
+	"time"
+
+	faiss "github.com/blevesearch/go-faiss"
+)
+
+type vecIndexCache struct {
+	closeCh chan struct{}
+	m       sync.RWMutex
+	cache   map[uint16]*cacheEntry
+}
+
+type ewma struct {
+	alpha  float64
+	avg    float64
+	sample uint64
+}
+
+type cacheEntry struct {
+	tracker *ewma
+	refs    int64
+	index   *faiss.IndexImpl
+}
+
+func newVectorIndexCache() *vecIndexCache {
+	return &vecIndexCache{
+		cache:   make(map[uint16]*cacheEntry),
+		closeCh: make(chan struct{}),
+	}
+}
+
+func (vc *vecIndexCache) Clear() {
+	vc.m.Lock()
+	close(vc.closeCh)
+	vc.cache = nil
```

Comment:
Think we should force close any open indexes at this point - asynchronously via a go routine.